File: frontend/src/api/social.ts
import { API_URL, getHeaders } from "./client";

export interface Friendship {
  id: number;
  from_user: string;
  to_user: string;
  status: 'pending' | 'accepted' | 'rejected' | 'blocked';
  created_at: string;
}

export interface FriendUser {
  id: number;
  username: string;
  email: string;
  first_name: string;
  last_name: string;
  phone_number: string;
  profile?: {
    avatar_config?: any;
    gender?: 'M' | 'F';
  };
}

export const sendFriendRequest = async (username: string): Promise<Friendship> => {
  const response = await fetch(`${API_URL}/users/friends/request`, {
    method: "POST",
    headers: getHeaders(),
    body: JSON.stringify({ username }),
  });

  if (!response.ok) {
    const err = await response.json();
    throw new Error(err.detail || "Failed to send friend request");
  }

  return response.json();
};

export const getPendingRequests = async (): Promise<Friendship[]> => {
  const response = await fetch(`${API_URL}/users/friends/requests/pending`, {
    method: "GET",
    headers: getHeaders(),
  });

  if (!response.ok) {
    throw new Error("Failed to fetch pending requests");
  }

  return response.json();
};

export const respondToFriendRequest = async (requestId: number, status: 'accepted' | 'rejected'): Promise<Friendship> => {
  const response = await fetch(`${API_URL}/users/friends/requests/${requestId}`, {
    method: "PATCH",
    headers: getHeaders(),
    body: JSON.stringify({ status }),
  });

  if (!response.ok) {
    const err = await response.json();
    throw new Error(err.detail || "Failed to respond to friend request");
  }

  return response.json();
};

export const getFriendsList = async (): Promise<FriendUser[]> => {
  const response = await fetch(`${API_URL}/users/friends`, {
    method: "GET",
    headers: getHeaders(),
  });

  if (!response.ok) {
    throw new Error("Failed to fetch friends list");
  }

  return response.json();
};

export const unfriendUser = async (username: string): Promise<void> => {
  const response = await fetch(`${API_URL}/users/friends/${username}`, {
    method: "DELETE",
    headers: getHeaders(),
  });

  if (!response.ok) {
    throw new Error("Failed to unfriend user");
  }
};

export const blockUser = async (username: string): Promise<Friendship> => {
  const response = await fetch(`${API_URL}/users/friends/block`, {
    method: "POST",
    headers: getHeaders(),
    body: JSON.stringify({ username }),
  });

  if (!response.ok) {
    const err = await response.json();
    throw new Error(err.detail || "Failed to block user");
  }

  return response.json();
};

File: backend/users/services/social.py
from django.contrib.auth import get_user_model
from django.db import models
from ..models import Friendship, FriendshipStatus

User = get_user_model()

def send_friend_request(*, from_user: User, to_username: str):
    to_user = User.objects.get(username=to_username)
    
    if from_user == to_user:
        raise ValueError("You cannot send a friend request to yourself.")
    
    existing = Friendship.objects.filter(
        (models.Q(from_user=from_user) & models.Q(to_user=to_user)) |
        (models.Q(from_user=to_user) & models.Q(to_user=from_user))
    ).first()
    
    if existing:
        if existing.status == FriendshipStatus.ACCEPTED:
            raise ValueError("You are already friends.")
        if existing.status == FriendshipStatus.PENDING:
            if existing.from_user == from_user:
                raise ValueError("Friend request already sent.")
            else:
                existing.status = FriendshipStatus.ACCEPTED
                existing.save()
                return existing
        if existing.status == FriendshipStatus.BLOCKED:
            if existing.from_user == from_user:
                raise ValueError("You have blocked this user.")
            else:
                raise ValueError("User has blocked you.")
        
        existing.from_user = from_user
        existing.to_user = to_user
        existing.status = FriendshipStatus.PENDING
        existing.save()
        return existing
    
    return Friendship.objects.create(from_user=from_user, to_user=to_user, status=FriendshipStatus.PENDING)

def respond_to_friend_request(*, friendship_id: int, user: User, status: str):
    if status not in [FriendshipStatus.ACCEPTED, FriendshipStatus.REJECTED]:
        raise ValueError("Invalid status response.")
        
    friendship = Friendship.objects.get(id=friendship_id)
    
    if friendship.to_user != user:
        raise ValueError("You can only respond to requests sent to you.")
        
    if friendship.status != FriendshipStatus.PENDING:
        raise ValueError("This request has already been processed.")
        
    friendship.status = status
    friendship.save()
    return friendship

def unfriend_user(*, user: User, target_username: str):
    target_user = User.objects.get(username=target_username)
    
    Friendship.objects.filter(
        (models.Q(from_user=user) & models.Q(to_user=target_user)) |
        (models.Q(from_user=target_user) & models.Q(to_user=user)),
        status=FriendshipStatus.ACCEPTED
    ).delete()

def block_user(*, user: User, target_username: str):
    target_user = User.objects.get(username=target_username)
    
    Friendship.objects.filter(
        (models.Q(from_user=user) & models.Q(to_user=target_user)) |
        (models.Q(from_user=target_user) & models.Q(to_user=user))
    ).delete()
    
    return Friendship.objects.create(from_user=user, to_user=target_user, status=FriendshipStatus.BLOCKED)
